[expo-updates][Android] Logging and log reading

douglowder opened the pull request:
# Why

Provides a way to log in expo-updates, and provides code to read all logs written since a given time.

# How

- New Kotlin code implementing the logger, log reader, expo-updates specific error codes, and JSON format for log messages
- Code is designed to match the error codes and log entry shape from the iOS implementation
- Logs with severity "error" or "fatal" will include stack traces (limited to 20 frames)
- Logs are written using the usual `android.util.Log` facility, with a tag specific to expo-updates. The log reader leverages `logcat` to read only system logs with our tag, and filters out logs that do not match the current process ID

# Test Plan

New unit tests provided for logging and log reading.

# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [x] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

douglowder committed:
[expo-updates][Android] Logging and log reading

Why

Provides a way to log in expo-updates, and provides code to read all logs written since a given time.

How

- New Kotlin code implementing the logger, log reader, expo-updates specific error codes, and JSON format for log messages
- Code is designed to match the error codes and log entry shape from the iOS implementation
- Logs with severity "error" or "fatal" will include stack traces
- Logs are written using the usual `android.util.Log` facility, with a tag specific to expo-updates. The log reader leverages `logcat` to read only system logs with our tag, and filters out logs that do not match the current process ID

Test Plan

New unit tests provided for logging and log reading.

wschurman commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:
```diff
@@ -0,0 +1,80 @@
+package expo.modules.updates
```
```suggestion
package expo.modules.updates.logging
```

wschurman commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:
```diff
@@ -0,0 +1,80 @@
+
+    val firstlogs = reader.getLogEntries(firstTime)
+    Assert.assertEquals(2, firstlogs.size)
+    Assert.assertEquals("Message 1", UpdatesLogEntry.create(firstlogs?.get(0)).message)
+    Assert.assertEquals("Message 2", UpdatesLogEntry.create(firstlogs?.get(1)).message)
```
```suggestion
    Assert.assertEquals("Message 1", UpdatesLogEntry.create(firstLogs[0]).message)
    Assert.assertEquals("Message 2", UpdatesLogEntry.create(firstLogs[1]).message)
```

wschurman commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:
```diff
@@ -0,0 +1,80 @@
+
+    val reader = UpdatesLogReader()
+
+    val firstlogs = reader.getLogEntries(firstTime)
```
```suggestion
    val firstLogs = reader.getLogEntries(firstTime)
```

wschurman commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:
```diff
@@ -0,0 +1,80 @@
+    Assert.assertEquals("Message 1", UpdatesLogEntry.create(firstlogs?.get(0)).message)
+    Assert.assertEquals("Message 2", UpdatesLogEntry.create(firstlogs?.get(1)).message)
+
+    val secondlogs = reader.getLogEntries(secondTime)
```
```suggestion
    val secondLogs = reader.getLogEntries(secondTime)
```

wschurman commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:
```diff
@@ -0,0 +1,80 @@
+
+    val secondlogs = reader.getLogEntries(secondTime)
+    Assert.assertEquals(1, secondlogs.size)
+    Assert.assertEquals("Message 2", UpdatesLogEntry.create(secondlogs?.get(0)).message)
+    Assert.assertEquals(MAX_FRAMES_IN_STACKTRACE, UpdatesLogEntry.create(secondlogs?.get(0)).stacktrace?.size)
```
```suggestion
    Assert.assertEquals("Message 2", UpdatesLogEntry.create(secondLogs.[0]).message)
    Assert.assertEquals(MAX_FRAMES_IN_STACKTRACE, UpdatesLogEntry.create(secondLogs[0]).stacktrace?.size)
```

wschurman commented on packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt:
```diff
@@ -0,0 +1,80 @@
+    val nowTimestamp = now.time / 1000
+    val sinceThen = Date(now.time - 5000)
+    val logs = UpdatesLogReader().getLogEntries(sinceThen)
+    var testPassed = false
+    for (log in logs) {
+      if (log.contains("{\"timestamp\":$nowTimestamp,\"message\":\"Test message\",\"code\":\"JSRuntimeError\",\"level\":\"warn\"}")) {
+        testPassed = true
+      }
+    }
+    Assert.assertTrue(testPassed)
```
```suggestion
    Assert.assertTrue(logs.any { it.contains("{\"timestamp\":$nowTimestamp,\"message\":\"Test message\",\"code\":\"JSRuntimeError\",\"level\":\"warn\"}") })
```

douglowder commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesErrorCode.kt:
```diff
@@ -0,0 +1,31 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
```
Good to know! I did see these headers in some iOS files...

douglowder replied:
Fixed

douglowder committed:
Update packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogEntry.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogEntry.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogReader.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogReader.kt:
```diff
@@ -0,0 +1,49 @@
+    val result: MutableList<String> = mutableListOf()
+    val epochTimestamp = newerThan.time / 1000
+    val pid = "${android.os.Process.myPid()}"
+    try {
+      // Use logcat to read just logs with our tag, in long format (message on separate line)
+      val process = Runtime.getRuntime().exec("logcat -d -s ${UpdatesLogger.LOGGING_TAG} -vlong")
+      val bufferedReader = BufferedReader(
+        InputStreamReader(process.inputStream)
+      )
+      var line: String? = ""
+      var writeThisLine = false
+      while (bufferedReader.readLine().also { line = it } != null) {
+        if (writeThisLine) {
+          // Check that it is a valid JSON string
+          val entry = UpdatesLogEntry.create(line ?: "")
+          // Check that timestamp is equal to or later than the passed in date before writing
+          if (entry?.timestamp >= epochTimestamp) {
+            result.add(line ?: "")
+          }
+          writeThisLine = false
+        }
+        if ((line ?: "").contains(pid)) {
+          // Line has our PID, so write the next line if needed
+          writeThisLine = true
+        }
+      }
+    } catch (e: IOException) {
+    }
+    return result
```
Interesting

douglowder replied:
This actually does not work, sadly. The logcat results come in sets of 3 lines, with an extra line at the beginning, except that just using chunk() needs care if only 0 or 1 entries are returned. So I'm going back to my original implementation.

douglowder replied:
I was able to get your nicer clean Kotlin code working to process the logcat output correctly, and have committed that along with explanatory comments.

douglowder committed:
Update packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogReader.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Update packages/expo-updates/android/src/androidTest/java/expo/modules/updates/logging/UpdatesLoggingTest.kt

Co-authored-by: Will Schurman <wschurman@gmail.com>

douglowder committed:
Address review comments

- Make adjustments and fix tests after committing suggestions
- Run :spotlessApply
- Back out use of chunk() etc. in UpdatesLogReader

douglowder commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesErrorCode.kt:
```diff
@@ -0,0 +1,31 @@
+// Copyright 2022-present 650 Industries. All rights reserved.
+
+// Error codes for expo-updates logs
```
Fixed

douglowder commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesErrorCode.kt:
```diff
@@ -0,0 +1,31 @@
+        JSRuntimeError -> "JSRuntimeError"
+      }
+    }
+  }
```
Fixed

douglowder commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogType.kt:
```diff
@@ -0,0 +1,47 @@
+import android.util.Log
+
+enum class UpdatesLogType(val type: Int) {
+  Trace(0),
```
Fixed

douglowder commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogType.kt:
```diff
@@ -0,0 +1,47 @@
+        Error -> "error"
+        Fatal -> "fatal"
+      }
+    }
```
Fixed

douglowder committed:
Clean up implementation of UpdatesLogReader

wschurman commented on packages/expo-updates/android/src/main/java/expo/modules/updates/logging/UpdatesLogReader.kt:
```diff
@@ -0,0 +1,39 @@
+  fun getLogEntries(newerThan: Date): List<String> {
+    val epochTimestamp = newerThan.time / 1000
+    val pid = "${android.os.Process.myPid()}"
+    try {
```
Awesome! Last thing on this one is that we probably want to not catch the error (just like in iOS).

douglowder replied:
Done

wschurman reviewed (approved):
Just one more change before commit.

douglowder committed:
Do not catch IOException in log reader

douglowder committed:
Update CHANGELOG.md
